perf(mangler)!: `Mangler::build_with_semantic` take mut ref to `Semantic`

overlookmotel opened the pull request:
`Semantic` is a very large type (600 bytes). The mangler only uses a part of it (`Scoping` and `AstNodes`), so it's a waste to copy the whole type onto the stack and then return it again.

Change `Mangler::build_with_semantic` to take `&mut Semantic` instead.

I hope this may also help a bit with wild variance in mangler benchmarks, as `Semantic` will no longer get dropped in the measured section (I don't think should be part of the measurement anyway, as the drop time of `Semantic` is not relevant in this benchmark).

overlookmotel commented:
* **#11132** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11132?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11132?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=11132).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented:
### Merge activity

* **May 18, 9:26 PM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **May 18, 9:26 PM EDT**: `Boshen` added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc).

overlookmotel committed:
perf(mangler)!: `Mangler::build_with_semantic` take mut ref to `Semantic` (#11132)

`Semantic` is a very large type (600 bytes). The mangler only uses a part of it (`Scoping` and `AstNodes`), so it's a waste to copy the whole type onto the stack and then return it again.

Change `Mangler::build_with_semantic` to take `&mut Semantic` instead.

I hope this may also help a bit with wild variance in mangler benchmarks, as `Semantic` will no longer get dropped in the measured section (I don't think should be part of the measurement anyway, as the drop time of `Semantic` is not relevant in this benchmark).
